File: Fortran/ex1/ex1.f90
	program ex1_AndrewNG
	implicit none
	
	integer,parameter :: m=97, iterations = 1500
	integer :: i,iter,nfile=20
	real,dimension(m) :: X, y
	real,dimension(2) :: theta = 0
	
	character(1024) :: buffer
  	character(20) :: var1, var2
  	integer :: pos

	real :: alpha = 0.01
	real :: JCost = 0
      	real :: temp0, temp1
	real :: predict1=0, predict2=0
	
	integer:: N=0, k, j
	real,dimension(3,3)::xx, yy, zz


	open(nfile, file='ex1data1.txt',status='old')
	do i = 1, m
		read(nfile,"(A)") buffer
  			pos = index(buffer, ",")
  	 		var1 = buffer(1:pos-1)
  		read(buffer(pos+1:), *) var2
		
		read(var1, * )X(i)
		read(var2, * )y(i)
	enddo
	close(nfile)
	
	open(nfile, file='ex1data1_2.txt',status='unknown')
	do i = 1, m
		write(nfile, * )X(i), y(i)
	enddo
	close(nfile)

	open(nfile,file='cmddata1.txt', status='replace')
!	write(nfile,'(a)')'symbol(z) = "o+"[int(z):int(z)]'
	write(nfile,'(a)')'plot "ex1data1_2.txt" '  ! using 1:2:(symbol($2+1)) with labels textcolor lt 1 \'	
	write(nfile,'(a)')'q'
	close(nfile) 		
	call system('gnuplot -persist cmddata1.txt')
	
	call costFunction(M, theta, X, y, Jcost) 
	write(*,*)
	write(*,*)'ans =', JCost
	write(*,*)
	
        write(*,*)'Now, lets try Batch Gradient Descent'
	call batchGradientDescent(M, alpha, iterations, X, y, theta)
	    
	write(*,*)'Batch Gradient Descent with iterations 1500'
	write(*,*)'theta(1)=', theta(1), 'theta(2)=', theta(2)
	
	write(*,*)
	predict1 = 10000 * theta(1) + 35000 * theta(2)
	write(*,*)'For population = 35,000, we predict a profit of ', predict1
	
	predict2 = 10000 * theta(1) + 70000 * theta(2)
	write(*,*)'For population = 70,000, we predict a profit of ', predict2
	
	write(*,*)
	write(*,*)'Now, lets try stochasitic Gradient Descent'
		
	write(*,*)'Stochastic Gradient Descent solutions are:'
	write(*,*)'theta(1)=', theta(1), 'theta(2)=', theta(2)
	write(*,*)
	
	
	write(*,*)
	write(*,*)'Now, lets try the closed form-direct way'	
	call LMS(X, y, m, theta)	
	write(*,*)'The closed form solutions are:'
	write(*,*)'theta(1)=', theta(1), 'theta(2)=', theta(2)

	
	end program

File: C/ex1/ex1.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
int N=2;

int LMS(double X[],double Y[],int M, double theta[]);
int transpose(int M, int N, double X[M][N], double XT[M][N]);
int multiply( int M, int N, int L, double X[M][N], double Y[N][L], double XML[M][L]);
void inverse(int dim, double mat[dim][dim], double matt[dim][dim] );

double costFunction(int M, double X[M],double Y[M], double theta[N]);
int batchGradientDescent(int M, int interations, double alpha,  double X[], double Y[], double theta[] );
int stochasiticGradientDescent(int M, double alpha,  double X[], double Y[], double theta[]);

int main( )
{
	int  i,iter,nfile=20;
	int M = 97;
	int iterations = 1500;
	
	double  X[M], Y[M];
	double  theta[2];

	double  alpha = 0.01;
	double  JCost = 0;
      double  temp0, temp1;
	double  predict1=0, predict2=0;	
	
	FILE *fp;
	char line[20];		
	
	if((fp = fopen ("ex1data1.txt", "r")) == NULL){
		printf("open wrong \n");
	}else{
	       i = 0;
	       while(fgets(line, sizeof line, fp) != NULL){
//		   printf("line[%d]=  %s \n", i, line); 
		   
		   X[i] = atof(strtok (line,","));		   	
    	         Y[i] = atof(strtok (NULL, ","));			
		   
//		   printf("X[%d]= %lf, Y[%d]= %lf \n", i, X[i], i,  Y[i]);
//             getchar();

		   i++;	   
	       }
	    fclose(fp);
	 }
	
	
	JCost = costFunction(M, X, Y, theta) ;    	
	printf("Cost Function(ans) = %f \n\n", JCost);
		
	
	
      printf("Now, lets try Batch Gradient Descent \n");	
      batchGradientDescent(M, iterations, alpha, X, Y, theta);
	printf("Batch Gradient Descent with iterations 1500 \n");
	printf("theta[0]= %f, theta[1]= %f  \n\n" , theta[0], theta[1]);
	

	
	
	predict1 = 10000 * theta[0] + 35000 * theta[1];
	printf("For population = 35,000, we predict a profit of  %f \n", predict1);
	
	predict2 = 10000 * theta[0] + 70000 * theta[1];
	printf("For population = 70,000, we predict a profit of %f \n\n ", predict2);
	
	printf("Now, lets try stochasitic Gradient Descent \n");
	stochasiticGradientDescent(M, alpha, X, Y, theta);	
	printf("Stochastic Gradient Descent solutions are: \n");
	printf("theta[0]= %f, theta[1]= %f \n\n" , theta[0], theta[1]);
	
	
	
	printf("Now, lets try the closed form \n ");	
		 if(LMS(X, Y, M, theta) != 0) printf("LMS wrong!");	
	printf("The closed form solutions are: \n ");
	printf("theta[0]= %f, theta[1]= %f \n\n" , theta[0], theta[1]);

	
      exit(0);
}	
	
